hash_practice/exercises.py

```diff
@@ -2,18 +2,61 @@
 def grouped_anagrams(strings):
     """ This method will return an array of arrays.
         Each subarray will have strings which are anagrams of each other
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(nlogn) - beacuse of the sort
+        Space Complexity: O(n)
     """
-    pass
+    repeated_words = {}
+    for word in strings:
+        sorted_word = "".join(sorted(word))
+        if sorted_word not in repeated_words:
+            repeated_words[sorted_word] = [word]
+        else:
+            repeated_words[sorted_word].append(word)
+    
+    anagrams_list = []
+    for words in repeated_words.values():
+        anagrams_list.append(words)
+    
+    return anagrams_list
+
+
 
 def top_k_frequent_elements(nums, k):
     """ This method will return the k most common elements
         In the case of a tie it will select the first occuring element.
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(n)
+        Space Complexity: O(n)
     """
-    pass
+    final_list = []
+    if len(nums) == 0:
+        return final_list
+    repeated_nums = {}
+    most_repeated = 0
+    for val in nums:
+        if val in repeated_nums:
+            repeated_nums[val] += 1
+        else:
+            repeated_nums[val] = 1
+        if repeated_nums[val] > most_repeated:
+            most_repeated = repeated_nums[val]
+    
+    order_repeated = {}
+    for key, val in repeated_nums.items():
+        if val in order_repeated:
+            order_repeated[val].append(key)
+        else:
+            order_repeated[val] = [key]
+        
+    while most_repeated > 0 and len(final_list) < k:
+        if most_repeated in order_repeated:
+            for num in order_repeated[most_repeated]:
+                final_list.append(num)
+        most_repeated -= 1
+    
+    return final_list
+
+
+
 
 
 def valid_sudoku(table):

```

Comment:
👍 Nice work, I would say the space complexity is also O(1) due to the board only being 9x9.  Great work!